src/environment.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use abstract_value::{self, AbstractValue, Path};
+use rpds::HashTrieMap;
+use rustc::hir::def_id::DefId;
+use rustc::ty::{self, Ty, TyCtxt};
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct Environment {
+    /// Does not include any entries where the value is abstract_value::Bottom
+    value_map: HashTrieMap<Path, AbstractValue>,
+}
+
+/// Constructor
+impl Environment {
+    /// Returns an environment that has a path for every parameter of the given function,
+    /// initialized with abstract_values::Top
+    pub fn with_parameters(def_id: DefId, tcx: TyCtxt) -> Environment {
+        let value_map = HashTrieMap::default();
+        let func_ty: Ty = tcx.type_of(def_id);
+        match func_ty.sty {
+            ty::FnDef(..) | ty::FnPtr(_) => {
+                let sig = func_ty.fn_sig(tcx);
+                let len = sig.inputs().skip_binder().len();
+                for i in 0..len {
+                    let par_i = Path::Parameter { index: i };
+                    // todo: figure out how to get a source span for each parameter definition
+                    value_map.insert(par_i, abstract_value::TOP);
+                }
+            }
+            _ => {}
+        };
+        Environment { value_map }
+    }
+}
+
+/// Methods
+impl Environment {
+    /// Returns a reference to the value associated with the given path.
+    /// If no such value exists, &abstract_value::Bottom is returned.
+    pub fn value_at(&self, path: &Path) -> &AbstractValue {
+        self.value_map.get(path).unwrap_or(&abstract_value::BOTTOM)
+    }
+
+    /// Returns an environment with a path for every entry in self and other and an associated
+    /// value that is the join of self.value_at(path) and other.value_at(path)
+    pub fn join(&self, other: &Environment, join_condition: &AbstractValue) -> Environment {
+        self.join_or_widen(other, join_condition, |x, y, c| x.join(y, c))
+    }
+
+    /// Returns an environment with a path for every entry in self and other and an associated
+    /// value that is the widen of self.value_at(path) and other.value_at(path)
+    pub fn widen(&self, other: &Environment, join_condition: &AbstractValue) -> Environment {
+        self.join_or_widen(other, join_condition, |x, y, c| x.widen(y, c))
+    }
+
+    /// Returns an environment with a path for every entry in self and other and an associated
+    /// value that is the join or widen of self.value_at(path) and other.value_at(path).
+    fn join_or_widen(
+        &self,
+        other: &Environment,
+        join_condition: &AbstractValue,
+        join_or_widen: fn(&AbstractValue, &AbstractValue, &AbstractValue) -> AbstractValue,
```

Comment:
The closure here feels a bit unfortunate, but I don't have a concrete suggestion for improvement.

Reply:
To be pedantic, join_or_widen is not a closure but a function pointer. I considered alternatives here and did not find them as readable. On the other hand, a function pointer introduces an indirection that will be hard for the compiler to get rid of (unless it aggressively inlines method join_or_widen into methods join and widen). That said, if an extra function call is going to be a performance problem here, a lot more rewriting will be required anyway, so I'll leave this as it is for now.